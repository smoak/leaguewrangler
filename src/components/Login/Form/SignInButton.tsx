import { Button } from '@mui/material';
import { FC } from 'react';

interface SignInButtonProps {
  readonly isDisabled: boolean;
  onClick(): void;
}

const SignInButton: FC<SignInButtonProps> = ({ isDisabled, onClick }) => {
  return (
    <Button
      type="submit"
      fullWidth
      variant="contained"
      onClick={(e) => {
        e.preventDefault();
        onClick();
      }}
      sx={{ mt: 3, mb: 2 }}
      disabled={isDisabled}
    >
      Sign in
    </Button>
  );
};

export default SignInButton;
