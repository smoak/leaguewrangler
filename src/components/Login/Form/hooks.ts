import { useCallback, useState } from 'react';

export interface ControlledInput {
  readonly value: string;
  onChange: React.ChangeEventHandler<HTMLTextAreaElement | HTMLInputElement>;
}
type UseControlledInput = (defaultValue: string) => ControlledInput;
export const useControlledInput: UseControlledInput = defaultValue => {
  const [value, setValue] = useState(defaultValue);
  const onChange = useCallback<React.ChangeEventHandler<HTMLTextAreaElement | HTMLInputElement>>(e => {
    setValue(e.target.value);
  }, []);

  return {
    value,
    onChange,
  };
};
