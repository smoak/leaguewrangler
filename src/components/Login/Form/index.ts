export { default } from './FormWithMutation';
export * from './Form';
