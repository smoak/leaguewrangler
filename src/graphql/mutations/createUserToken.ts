import { gql } from 'apollo-boost';

export default gql`
  mutation CreateUserToken($username: String!, $password: String!) {
    createUserToken(username: $username, password: $password) {
      token
    }
  }
`;
