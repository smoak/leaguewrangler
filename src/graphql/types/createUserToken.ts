import * as Types from './globalTypes';

export type CreateUserTokenMutationVariables = Types.Exact<{
  username: Types.Scalars['String'];
  password: Types.Scalars['String'];
}>;


export type CreateUserTokenMutation = { __typename?: 'Mutation', createUserToken?: { __typename?: 'UserToken', token: string } | null };
