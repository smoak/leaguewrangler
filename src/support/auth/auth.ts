const userTokenName = 'userToken';

export type GetUserToken = () => string | null;
export const getUserToken: GetUserToken = () => {
  return localStorage.getItem(userTokenName);
};

export type SetUserToken = (token: string) => void;
export const setUserToken: SetUserToken = token => localStorage.setItem(userTokenName, token);
