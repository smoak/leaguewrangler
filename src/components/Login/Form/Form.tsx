import { useMutation } from '@apollo/client';
import { Checkbox, CircularProgress, FormControl, FormControlLabel, Input, InputLabel } from '@material-ui/core';
import { FC, useCallback } from 'react';
import { setUserToken } from 'support/auth';

import CREATE_USER_TOKEN from '../../../graphql/mutations/createUserToken';
import { CreateUserToken, CreateUserTokenVariables } from '../../../graphql/types/CreateUserToken';

import SignInButton from './SignInButton';
import { useControlledInput } from './hooks';

export interface FormProps {
  readonly formClassname: string;
  readonly submitButtonClassname: string;
}

const Form: FC<FormProps> = ({ submitButtonClassname, formClassname }) => {
  const { value: username, onChange: onUsernameChange } = useControlledInput('');
  const { value: password, onChange: onPasswordChange } = useControlledInput('');
  const [createUserToken, { data, error, loading }] = useMutation<CreateUserToken, CreateUserTokenVariables>(
    CREATE_USER_TOKEN
  );

  const onSignInClicked = useCallback(() => {
    createUserToken({ variables: { username, password } });
  }, [createUserToken, username, password]);

  if (data && data.createUserToken) {
    setUserToken(data.createUserToken.token);
    window.location.reload();
  }

  const button = loading ? (
    <CircularProgress color="secondary" />
  ) : (
    <SignInButton onClick={onSignInClicked} isDisabled={!username || !password} className={submitButtonClassname} />
  );

  return (
    <div className={formClassname}>
      <FormControl margin="normal" required fullWidth error={!!error}>
        <InputLabel htmlFor="username">Username</InputLabel>
        <Input
          id="username"
          name="username"
          autoComplete="username"
          autoFocus={true}
          value={username}
          onChange={onUsernameChange}
        />
      </FormControl>
      <FormControl margin="normal" required fullWidth error={!!error}>
        <InputLabel htmlFor="password">Password</InputLabel>
        <Input
          name="password"
          type="password"
          id="password"
          autoComplete="current-password"
          value={password}
          onChange={onPasswordChange}
          data-testid="password-textbox"
        />
      </FormControl>
      <FormControlLabel control={<Checkbox value="remember" color="primary" />} label="Remember me" />
      {button}
    </div>
  );
};

export default Form;
