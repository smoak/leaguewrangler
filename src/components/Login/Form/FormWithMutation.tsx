import React, { FC } from 'react';
import { Omit } from '@material-ui/core';

import Form, { FormProps } from './Form';
import { Mutation, MutationOpts } from 'react-apollo';
import { CreateUserToken, CreateUserTokenVariables } from '../../../graphql/types/CreateUserToken';
import createUserToken from '../../../graphql/mutations/createUserToken';
import { setUserToken } from '../../../support/auth';

export type FormWithMutationProps = Omit<FormProps, 'onSignInClicked'>;

export class SignInMutation extends Mutation<CreateUserToken, CreateUserTokenVariables> {}

const onSignInCompleted: MutationOpts<CreateUserToken, CreateUserTokenVariables>['onCompleted'] = ({
  createUserToken: { token },
}) => {
  setUserToken(token);
  window.location.reload();
};

const FormWithMutation: FC<FormWithMutationProps> = props => {
  return (
    <SignInMutation mutation={createUserToken} onCompleted={onSignInCompleted}>
      {mutate => (
        <Form
          {...props}
          onSignInClicked={(username, password) => {
            mutate({
              variables: { username, password },
            });
          }}
        />
      )}
    </SignInMutation>
  );
};

export default FormWithMutation;
