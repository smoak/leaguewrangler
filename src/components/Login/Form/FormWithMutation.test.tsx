jest.mock('../../../support/auth');

import React from 'react';
import FormWithMutation, { FormWithMutationProps, SignInMutation } from './FormWithMutation';
import { ShallowWrapper, shallow, ReactWrapper, mount } from 'enzyme';
import { MockedProvider, MockedResponse } from 'react-apollo/test-utils';
import wait from 'waait';

import { setUserToken } from '../../../support/auth';
import Form from './Form';
import createUserToken from '../../../graphql/mutations/createUserToken';

describe('FormWithMutation', () => {
  let component: ShallowWrapper<FormWithMutationProps>;

  beforeEach(() => {
    component = shallow(<FormWithMutation formClassname="form" submitButtonClassname="submit" />);
  });

  it('renders', () => {
    expect(component).toMatchSnapshot();
  });

  describe('when the form sign in callback is executed', () => {
    let wrappedComponent: ReactWrapper;

    beforeEach(async () => {
      Object.defineProperty(window.location, 'reload', {
        configurable: true,
      });
      window.location.reload = jest.fn();

      const mocks: ReadonlyArray<MockedResponse> = [
        {
          request: {
            query: createUserToken,
            variables: {
              username: 'foo',
              password: 'bar',
            },
          },
          result: {
            data: {
              createUserToken: {
                token: 'foo',
              },
            },
          },
        },
      ];
      wrappedComponent = mount(
        <MockedProvider mocks={mocks} addTypename={false}>
          <FormWithMutation formClassname="form" submitButtonClassname="submit" />
        </MockedProvider>
      );

      wrappedComponent.find(Form).prop('onSignInClicked')('foo', 'bar');

      await wait(0);
    });

    it('sets the users token', () => {
      expect(setUserToken).toHaveBeenCalledWith('foo');
    });

    it('reloads the page', () => {
      expect(window.location.reload).toHaveBeenCalled();
    });
  });

  describe('when the mutation completes', () => {
    beforeEach(async () => {
      Object.defineProperty(window.location, 'reload', {
        configurable: true,
      });
      window.location.reload = jest.fn();

      component = shallow(<FormWithMutation formClassname="form" submitButtonClassname="submit" />);
      (component.find(SignInMutation).prop('onCompleted') as SignInMutation['onMutationCompleted'])({
        createUserToken: { token: 'foo' },
      });
    });

    it('sets the users token', () => {
      expect(setUserToken).toHaveBeenCalledWith('foo');
    });

    it('reloads the page', () => {
      expect(window.location.reload).toHaveBeenCalled();
    });
  });
});
