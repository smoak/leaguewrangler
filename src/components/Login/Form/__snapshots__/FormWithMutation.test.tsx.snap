// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormWithMutation renders 1`] = `
<SignInMutation
  mutation={
    Object {
      "definitions": Array [
        Object {
          "directives": Array [],
          "kind": "OperationDefinition",
          "name": Object {
            "kind": "Name",
            "value": "CreateUserToken",
          },
          "operation": "mutation",
          "selectionSet": Object {
            "kind": "SelectionSet",
            "selections": Array [
              Object {
                "alias": undefined,
                "arguments": Array [
                  Object {
                    "kind": "Argument",
                    "name": Object {
                      "kind": "Name",
                      "value": "username",
                    },
                    "value": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "username",
                      },
                    },
                  },
                  Object {
                    "kind": "Argument",
                    "name": Object {
                      "kind": "Name",
                      "value": "password",
                    },
                    "value": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "password",
                      },
                    },
                  },
                ],
                "directives": Array [],
                "kind": "Field",
                "name": Object {
                  "kind": "Name",
                  "value": "createUserToken",
                },
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "token",
                      },
                      "selectionSet": undefined,
                    },
                  ],
                },
              },
            ],
          },
          "variableDefinitions": Array [
            Object {
              "defaultValue": undefined,
              "directives": Array [],
              "kind": "VariableDefinition",
              "type": Object {
                "kind": "NonNullType",
                "type": Object {
                  "kind": "NamedType",
                  "name": Object {
                    "kind": "Name",
                    "value": "String",
                  },
                },
              },
              "variable": Object {
                "kind": "Variable",
                "name": Object {
                  "kind": "Name",
                  "value": "username",
                },
              },
            },
            Object {
              "defaultValue": undefined,
              "directives": Array [],
              "kind": "VariableDefinition",
              "type": Object {
                "kind": "NonNullType",
                "type": Object {
                  "kind": "NamedType",
                  "name": Object {
                    "kind": "Name",
                    "value": "String",
                  },
                },
              },
              "variable": Object {
                "kind": "Variable",
                "name": Object {
                  "kind": "Name",
                  "value": "password",
                },
              },
            },
          ],
        },
      ],
      "kind": "Document",
      "loc": Object {
        "end": 156,
        "start": 0,
      },
    }
  }
  onCompleted={[Function]}
>
  <Component />
</SignInMutation>
`;
